UPGRADE.md

```diff
@@ -35,6 +35,191 @@ starts new workers, which take over from old workers before those old workers
+## Upgrade to 3.0.x
+
+Kong adheres to [semantic versioning](https://semver.org/), which makes a
+distinction between "major", "minor", and "patch" versions. The upgrade path
+will be different depending on which previous version from which you are migrating.
+
+Upgrading into 3.0.x is a major upgrade,
+be aware of any [breaking changes](https://github.com/Kong/kong/blob/release/3.0.x/CHANGELOG.md#breaking-changes)
+between the between 2.x and 3.x series. For 1.x series, also referer to
+[breaking changes of 2.x](#breaking-changes-2.0.0).
+
+### Dependencies
+
+If you are using the provided binary packages, all necessary dependencies
+for the gateway are bundled and you can skip this section.
+
+If you are building your dependencies by hand, there are changes since the
+previous release, so you will need to rebuild them with the latest patches.
+
+The required version of OpenResty is bumped up to [1.21.4.1](https://openresty.org/en/ann-1021004001.html).
+We recommend you to use the [openresty-build-tools](https://github.com/Kong/kong-build-tools/tree/master/openresty-build-tools),
+which allows you to more easily build OpenResty with the necessary patches and modules.
+
+
+### Template changes
+
+There are **Changes in the Nginx configuration file**, between kong 2.0.x,
+2.1.x, 2.2.x, 2.3.x, 2.4.x, 2.5.x, 2.6.x, 2.7.x, 2.8.x and 3.0.x
+
+To view the configuration changes between versions, clone the
+[Kong repository](https://github.com/kong/kong) and run `git diff`
+on the configuration templates, using `-w` for greater readability.
+
+Here's how to see the differences between previous versions and 3.0.x:
+
+```
+git clone https://github.com/kong/kong
+cd kong
+git diff -w 2.0.0 3.0.0 kong/templates/nginx_kong*.lua
+```
+
+**Note:** Adjust the starting version number
+(2.0.x, 2.1.x, 2.2.x, 2.3.x, 2.4.x, 2.5.x, 2.6.x, 2.7.x) to the version number you are currently using.
+
+To produce a patch file, use the following command:
+
+```
+git diff 2.0.0 3.0.0 kong/templates/nginx_kong*.lua > kong_config_changes.diff
+```
+
+**Note:** Adjust the starting version number
+(2.0.x, 2.1.x, 2.2.x, 2.3.x, 2.4.x, 2.5.x, 2.6.x, 2.7.x) to the version number you are currently using.
+
+
+### Suggested upgrade path
+
+**Version prerequisites for migrating to version 3.0.x**
+
+Kong supports a no-downtime migration model since 1.0. This means that while the
+migration is ongoing, you will have two Kong clusters running, sharing the
+same database. (This is sometimes called the Blue/Green migration model.)
+
+The lowest version that Kong 3.0.x supports migrating from with no downtime is 2.8.x for hybrid mode,
+and 2.8.2 or beyond for traditional mode.
+
+To upgrade to 2.8.x first, please follow the steps described in the section
+[Upgrade from 1.0.x - 2.2.x to 3.0.x](#upgrade-from-10x---22x-to-28x)
+
+Please be aware that you should not change configuration in the process of the blue-green migration, as it may cause unexpected behaviors and break your configuration.
```

Comment:
Should we change this phrase `blue-green migration`?

Reply:
This line is too long.